libs/vectorise/include/vectorise/uint/int.hpp

```diff
@@ -0,0 +1,1298 @@
+  std::size_t nbits       = WIDE_ELEMENT_SIZE - real_bits;
+  // No actual shifting involved, just move the elements
+  if (full_words != 0u)
+  {
+    for (std::size_t i = 0; i < WIDE_ELEMENTS - full_words; i++)
+    {
+      wide_[i] = wide_[i + full_words];
+    }
+    for (std::size_t i = 0; i < full_words; i++)
+    {
+      wide_[WIDE_ELEMENTS - i - 1] = mask;
+    }
+  }
+
+  if (is_positive)
+  {
+    // If real_bits == 0, nothing to do
+    if (real_bits != 0u)
+    {
+      WideType carry = 0;
+      for (std::size_t i = 0; i < WIDE_ELEMENTS; i++)
+      {
+        WideType val                 = wide_[WIDE_ELEMENTS - 1 - i];
+        wide_[WIDE_ELEMENTS - 1 - i] = (val >> real_bits) | carry;
+        carry                        = val << nbits;
+      }
+    }
+  }
+
+  return *this;
+}
+
+template <uint16_t S>
+constexpr Int<S> Int<S>::Sign() const
+{
+  return IsPositive() ? Int<S>::_1 : -(Int<S>::_1);
+}
+
+template <uint16_t S>
+constexpr bool Int<S>::IsPositive() const
+{
+  return ((wide_[WIDE_ELEMENTS - 1] & 0x8000000000000000ULL) >> 63) == 0;
+}
+
+template <uint16_t S>
+constexpr std::size_t Int<S>::msb() const
+{
+  std::size_t msb = 0;
+  for (std::size_t i = 0; i < WIDE_ELEMENTS; i++)
+  {
+    std::size_t msbi = platform::CountLeadingZeroes64(wide_[WIDE_ELEMENTS - 1 - i]);
+    msb += msbi;
+    if (msbi < 64)
+    {
+      return msb;
+    }
+  }
+  return msb;
+}
+
+template <uint16_t S>
+constexpr std::size_t Int<S>::lsb() const
+{
+  std::size_t lsb = 0;
+  for (std::size_t i = 0; i < WIDE_ELEMENTS; i++)
+  {
+    std::size_t lsbi = platform::CountTrailingZeroes64(wide_[i]);
+    lsb += lsbi;
+    if (lsbi < 64)
+    {
+      return lsb;
+    }
+  }
+  return lsb;
+}
+
+template <uint16_t S>
+constexpr uint64_t Int<S>::TrimmedSize() const
+{
+  uint64_t ret = WIDE_ELEMENTS;
+  while ((ret != 0) && (wide_[ret - 1] == 0))
+  {
+    --ret;
+  }
+  return ret;
+}
+
+template <uint16_t S>
+constexpr uint64_t Int<S>::size() const
+{
+  return ELEMENTS;
+}
+
+template <uint16_t S>
+constexpr uint64_t Int<S>::elements() const
+{
+  return WIDE_ELEMENTS;
+}
+
+template <uint16_t S>
+inline std::ostream &operator<<(std::ostream &s, Int<S> const &x)
+{
+  s << std::string(x);
+  return s;
+}
+
+}  // namespace vectorise
+
+namespace serializers {
+template <typename D, uint16_t S>
+struct ArraySerializer<vectorise::Int<S>, D>
+{
+public:
+  using Type       = vectorise::Int<S>;
+  using DriverType = D;
+
+  template <typename Constructor>
+  static void Serialize(Constructor &array_constructor, Type const &u)
+  {
+    // TODO(issue 1425): Add WideType size to serialisation
+    auto array = array_constructor(u.elements());
+    for (std::size_t i = 0; i < u.elements(); i++)
+    {
+      array.Append(u.ElementAt(i));
+    }
+  }
+
+  template <typename ArrayDeserializer>
+  static void Deserialize(ArrayDeserializer &array, Type &u)
+  {
+    if (array.size() != u.elements())
+    {
+      // TODO(?): Throw
```

Comment:
throwing sounds sensible to me. you may want to make a custom exception that explains what happened here.

Reply:
this is also the case in the existing UInt class -where this code was copied from. I will investigate further

Reply:
I would like to tackle this in a separate issue and not as part of this PR.